src/collapse/perf.rs

```diff
@@ -324,6 +387,91 @@ impl PerfState {
         }
         Ok(())
     }
+
+    // Use addr2line to determine the symbol to use for each program counter address
+    // (as opposed to using the symbol names that perf script produces)
+    fn un_inline(&mut self, addr: &str, module: &str) -> result::Result<String, Error> {
+        let addr = u64::from_str_radix(addr, 16).map_err(|e| {
+            Error::Parse(format!(
+                "unable to parse {} as hex address, caused by: {}",
+                addr, e
+            ))
+        })?;
+
+        if let Entry::Vacant(entry) = self.addr2line_contexts.entry(module.into()) {
+            let bytes = fs::read(module)?;
+            let file = &object::File::parse(&bytes).map_err(|e| {
+                Error::Parse(format!(
+                    "unable to parse file: {}, caused by: {}",
+                    module, e
+                ))
+            })?;
+            entry.insert(addr2line::Context::new(file)?);
+        };
+        let ctx = &self.addr2line_contexts[module];
+
+        let mut frames = ctx.find_frames(addr).map_err(|e| {
+            Error::Parse(format!(
+                "unable to parse frames from module: {}, at address: {}. caused by: {}",
+                module, addr, e
+            ))
+        })?;
+
+        let mut full_func = SmallVec::<[String; 1]>::new();
+        let mut total_length = 0;
+        while let Some(frame) = frames.next().unwrap() {
+            let mut function = frame
+                .function
+                .map(|func| {
+                    let name = remove_discriminator(
+                        func.raw_name()
+                            .map(Cow::from)
+                            .unwrap_or_else(|_| Cow::from("??")),
+                    );
+                    strip_rust_hash(addr2line::demangle_auto(name, func.language).as_ref())
+                        .to_string()
```

Comment:
It seems unfortunate for this code to throw away the `Cow<str>` with `.as_ref()` for `strip_rust_hash`, and then to turn it back into a `String` here, when the `Cow<str>` _could_ already be owned.

Reply:
I'm removing strip_rust_hash so there's no longer a need to change it.